pkg/migration-tool/planner/planner.go

```diff
@@ -0,0 +1,239 @@
+	reference = &Block{
+		id:                    id,
+		percent:               percent,
+		pbarDescriptionPrefix: baseDescription,
+		pbar: progressbar.NewOptions(
+			6,
+			progressbar.OptionOnCompletion(func() {
+				_, _ = fmt.Fprint(os.Stderr, "\n")
+			}),
+		),
+		mint:    mint,
+		maxt:    maxt,
+		pbarMux: p.pbarMux,
+		plan:    p,
+	}
+	if p.Quite {
+		reference.pbar = nil
+		reference.pbarDescriptionPrefix = ""
+	}
+	if p.TestCheckFunc != nil {
+		// This runs only during integration tests. It enables the tests to access and test the internal
+		// state of the planner.
+		p.TestCheckFunc()
+	}
+	reference.SetDescription(fmt.Sprintf("fetching time-range: %d mins...", timeRangeInMinutes), 1)
+	return
+}
+
+func (p *Plan) validateT(mint, maxt int64) error {
+	switch {
+	case p.Mint > mint || p.Maxt < mint:
+		return fmt.Errorf("invalid mint: %d: global-mint: %d and global-maxt: %d", mint, p.Mint, p.Maxt)
+	case p.Mint > maxt || p.Maxt < maxt:
+		return fmt.Errorf("invalid maxt: %d: global-mint: %d and global-maxt: %d", mint, p.Mint, p.Maxt)
+	case mint > maxt:
+		return fmt.Errorf("mint cannot be greater than maxt: mint: %d and maxt: %d", mint, maxt)
+	}
+	return nil
+}
+
+func determineTimeDelta(numBytes, limit, deltaIncRegion int64, prevTimeDelta int64) int64 {
+	switch {
+	case numBytes <= deltaIncRegion:
+		// We increase the time-range linearly for the next fetch if the current time-range fetch resulted in size that is
+		// less than half the aimed maximum size of the in-memory block. This continues till we reach the
+		// maximum time-range delta.
+		if prevTimeDelta >= maxTimeRangeDeltaLimit {
+			return maxTimeRangeDeltaLimit
+		}
+		return clampTimeDelta(prevTimeDelta, laIncrement)
+	case numBytes > limit:
+		// The priority of size is more than that of time. That means, even if both bytes and time are greater than
+		// the respective limits, then we down size the time so that bytes size can be controlled (preventing OOM).
+		log.Info("msg", fmt.Sprintf("decreasing time-range delta to %d minute(s) since size beyond permittable limits", prevTimeDelta/(2*minute)))
+		return prevTimeDelta / 2
+	}
+	// Here, the numBytes is between the deltaIncRegion and the limit (i.e., deltaIncRegion < numBytes <= limit).
+	// This region is an ideal case of balance between migration speed and memory utilization.
+	// Example: If the limit is 500MB, then the deltaIncRegion will be 250MB. This means that till the numBytes is below
+	// 250MB, the time-range for next fetch will by increased by 1 minute (on the previous fetch time-range). However,
+	// the moment any block comes between 250MB and 500MB, we stop to increment the time-range delta further. This keeps
+	// the migration tool in very safe memory limits.
+	return prevTimeDelta
+}
+
+func clampTimeDelta(t int64, incr int64) int64 {
+	t += incr
+	if t > maxTimeRangeDeltaLimit {
+		return maxTimeRangeDeltaLimit
+	}
+	return t
+}
```

Comment:
nit : a clamp function is traditionally written as the following and the math is written on input
```suggestion
func clampTimeDelta(t int64) int64 {
	if t > maxTimeRangeDeltaLimit {
		return maxTimeRangeDeltaLimit
	}
	return t
}
```